Sources/Datasets/datasets/image/MNIST.swift

```diff
@@ -1,99 +1,108 @@
-    }
-}
+    let split: S5TFSplit
+    let batchSize: Int
+    let data: Tensor<UInt8>
+    let labels: Tensor<Int64>
 
-public struct MNIST: S5TFDataset {
-    typealias _Iterator = MNISTIterator // swiftlint:disable:this type_name
-
-    internal var data = [[1, 2], [3, 4], [5, 6]]
-
-    public func batched(_ batchSize: Int) -> MNISTIterator {
-        return MNISTIterator(dataset: self)
-    }
-
-    public var train: MNIST {
-        guard self.split == .undefined else { fatalError("This property can only be accessed from an undefined split.") }
-        return MNIST(split: .train)
-    }
-    public var validation: MNIST {
-        fatalError("The split does not exist for this dataset.")
-    }
-    public var test: MNIST {
-        guard self.split == .undefined else { fatalError("This property can only be accessed from an undefined split.") }
-        return MNIST(split: .test)
-    }
-    public var all: MNIST {
-        guard self.split == .undefined else { fatalError("This property can only be accessed from an undefined split.") }
-        return MNIST(split: .all)
-    }
-
-    public let split: S5TFSplit
-
-    public init() { self.split = .undefined }
-    private init(split: S5TFSplit) {
+    internal init(split: S5TFSplit, batchSize: Int) {
+        self.batchSize = batchSize
         self.split = split
-        let downloader = Downloader()
+
+        let semaphore = DispatchSemaphore(value: 0)
         let baseURL = "https://storage.googleapis.com/cvdf-datasets/mnist"
-        func getTrain() {
-            downloader.download(fileAt: URL(string: baseURL + "/train-images-idx3-ubyte.gz")!,
-            					cacheName: "mnist", fileName: "mnist_train_images") { url, error in }
-            downloader.download(fileAt: URL(string: baseURL + "/train-labels-idx1-ubyte.gz")!,
-            					cacheName: "mnist", fileName: "mnist_train_labels") { url, error in }
-        }
-        func getTest() {
-            downloader.download(fileAt: URL(string: baseURL + "/t10k-images-idx3-ubyte.gz")!,
-            					cacheName: "mnist", fileName: "mnist_test_images") { url, error in }
-            downloader.download(fileAt: URL(string: baseURL + "/t10k-labels-idx1-ubyte.gz")!,
-            					cacheName: "mnist", fileName: "mnist_test_labels") { url, error in }
-        }
+        var dataURL: URL? = nil
+        var labelURL: URL? = nil
+
+        print("Fetching files... Waiting for the download to finish before continuing...")
         switch split {
             case .train:
-                getTrain()
+                Downloader().download(fileAt: URL(string: baseURL + "/train-images-idx3-ubyte.gz")!,
+                                      cacheName: "mnist", fileName: "mnist_train_images") { url, err in
+                                      guard let url = url else {
+                                      fatalError("Data not downloaded.")
+                                      }
```

Comment:
Consider changing the indentation here.

In [`Iris.swift`](https://github.com/rickwierenga/datasets-1/blob/iris/Sources/Datasets/datasets/structured/Iris.swift) I added a little check to see whether we have access to the error, and print it out if we do:

```swift
cacheName: "iris", fileName: "iris.csv") { url, error in
     guard let url = url else {
         fatalError("URL not saved.")
         if let error = error { print(error) }
         fatalError("Data not could not be downloaded.")
     }
     ...
}
```